// ngx-loading-bar/LoadingBarService.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var core_1 = require("@angular/core");
var LoadingBarService = (function () {
    function LoadingBarService() {
        this.onStart = new core_1.EventEmitter();
        this.onStop = new core_1.EventEmitter();
        this.onComplete = new core_1.EventEmitter();
        this.onReset = new core_1.EventEmitter();
    }
    LoadingBarService.prototype.start = function () {
        this.onStart.emit(undefined);
    };
    LoadingBarService.prototype.stop = function () {
        this.onStop.emit(undefined);
    };
    LoadingBarService.prototype.complete = function () {
        this.onComplete.emit(undefined);
    };
    LoadingBarService.prototype.reset = function () {
        this.onReset.emit(undefined);
    };
    return LoadingBarService;
}());
LoadingBarService = __decorate([
    core_1.Injectable()
], LoadingBarService);
exports.LoadingBarService = LoadingBarService;
