// angular-progress-bar/src/progress-bar/progress-bar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var ProgressBarComponent = (function () {
    function ProgressBarComponent() {
        this.color = "#488aff";
    }
    ProgressBarComponent.prototype.whichColor = function (percent) {
        var k = Object.keys(this.degraded);
        k.forEach(function (e, i) { return k[i] = +e; });
        k = k.sort(function (a, b) { return a - b; });
        var p = +percent;
        var last = k[0];
        for (var _i = 0, k_1 = k; _i < k_1.length; _i++) {
            var val = k_1[_i];
            if (val < p) {
                last = val;
            }
            else if (val >= p - 1) {
                return this.degraded[last];
            }
        }
        return this.degraded[last];
    };
    ProgressBarComponent.prototype.whichProgress = function (progress) {
        try {
            return Math.round(progress * 100) / 100;
        }
        catch (_a) {
            return progress;
        }
    };
    ProgressBarComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'progress-bar',
                    styles: ["\n          .progress-outer {\n            width: 96%;\n            margin: 10px 2%;\n            padding: 3px;\n            text-align: center;\n            background-color: #f4f4f4;\n            border: 1px solid #dcdcdc;\n            color: #fff;\n            border-radius: 20px;\n          }\n          .progress-inner {\n            min-width: 15%;\n            white-space: nowrap;\n            overflow: hidden;\n            padding: 5px;\n            border-radius: 20px;\n          }\n    "],
                    template: "<div class=\"progress-outer\">\n" +
                        "    <div class=\"progress-inner\" [style.width]=\"whichProgress(progress) + '%'\" [style.background-color]=\"degraded == null ? color : whichColor(progress)\">\n" +
                        "        {{whichProgress(progress)}}%\n" +
                        "</div>\n" +
                        "</div>"
                },] },
    ];
    ProgressBarComponent.ctorParameters = function () { return []; };
    ProgressBarComponent.propDecorators = {
        'progress': [{ type: core_1.Input, args: ['progress',] },],
        'color': [{ type: core_1.Input, args: ['color',] },],
        'degraded': [{ type: core_1.Input, args: ['color-degraded',] },],
    };
    return ProgressBarComponent;
}());
exports.ProgressBarComponent = ProgressBarComponent;
